teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: False
+    reason: ExampleFailed
+    message: Oh no bad times
+```
+
+The `Run` type's `.status` will also allow controllers to report other fields, such as `startTime`, `completionTime`, `results` (see below), and arbitrary context-dependent fields the Custom Task author wants to report. A fully-specified `Run` status might look like:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+  completionTime: "2020-06-18T11:55:01Z"
+  startTime: "2020-06-18T11:55:01Z"
+  results:
+  - name: first-name
+    value: Bob
+  - name: last-name
+    value: Smith
+  arbitraryField: hello world
+  arbitraryStructuredField:
+    listOfThings: ["a", "b", "c"] 
+```
+
+### Pipeline Integration
+
+Enabling `Run`s by themselves are not terribly compelling. Their power comes from being specified in `Pipeline`s, and executed during `PipelineRun`s.
+
+Under this proposal, a user can define a Pipeline that invokes a Custom Task, specified similar to how `Task`s are specified today:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example-pipeline
+spec:
+  tasks:
+  …
+  - name: example-task
+    taskRef:
+      apiVersion: example.dev/v0
+      kind: Example
+      name: my-example
+```
+
+When a `PipelineRun` is created referencing this `Pipeline` definition, when `example-task` is next to execute, Tekton's `PipelineRun` controller will create a `Run` referencing the `Example` object, at which point it the Custom Task author's controller will be notified and it will be that controller's responsibility to take some action and eventually update the `Run`'s `.status` to signal completion.
+
+The `PipelineRun` controller will watch `Run` objects it's created, and take appropriate action when they report success or failure.
+
+### Initial Update Timeout
+
+It's possible that, though the Custom Task author has defined their CRD type, there's no controller watching for `Run`s of that type and updating their statuses as expected (e.g., it's crash-looping, it has been uninstalled, or it was never provided).
+
+In this case, a `PipelineRun` that depends on that type may wait uselessly for the execution to complete, until its configured timeout, which might be hours later. To save users' time and fail fast, the `PipelineRun`  controller will enforce a short timeout for initial updates to `Run` objects. If a `Run` hasn't been updated to the condition `.status.conditions[@type=Succeeded]=Unknown` after a certain (configurable) amount of time, say 30 seconds, then the `PipelineRun`  controller should fail the `PipelineRun` with a descriptive error message. This simple update indicates that the execution is running, and at least something is consuming new `Run` creations and updating statuses.
```

Comment:
You mean how we deal with the Broker / Trigger that does not have an owner? For Trigger where can see that there's no corresponding Broker we can do this in a sweep. For, Broker since we control this via brokerclass annotation, we wanted to use the defaulting webhook to mark the CR as 'unowned', but this does not work. If we use v1 CRD, we can default the status there.
So, I think when the custom CRD is created, it might be a good idea to ask them to use the v1 crd and specify that they should default the status.conditions@ in such a way as to indicate that it's unowned, guess leaving it empty could suffice to indicate that, but just a thought.